### PR TITLE
[Enhancement] Time log wages and multiple active salary structure

### PR DESCRIPTION
**Pull request by rohitwaghchaure**

### CODE REVIEW

File: erpnext/accounts/doctype/accounts_settings/accounts_settings.json [hunk 1]
```diff
@@ -130,19 +136,47 @@
...
    "set_only_once": 0, 
    "unique": 0
+  }, 
+  {
+   "allow_on_submit": 0, 
+   "bold": 0, 
+   "collapsible": 0, 
+   "default": "1", 
+   "description": "If enabled, the user can make sales invoice from employee's timesheet.", 
+   "fieldname": "time_sheet_for_sales_invoice",
```
Comment (rmehta) [inline erpnext/accounts/doctype/accounts_settings/accounts_settings.json]: remove description

File: erpnext/accounts/doctype/accounts_settings/accounts_settings.py [hunk 2]
```diff
@@ -12,6 +12,7 @@
 class AccountsSettings(Document):
 	def on_update(self):
 		frappe.db.set_default("auto_accounting_for_stock", self.auto_accounting_for_stock)
+		frappe.db.set_default("time_sheet_for_sales_invoice", self.time_sheet_for_sales_invoice)
```
Comment (rmehta) [inline erpnext/accounts/doctype/accounts_settings/accounts_settings.py]: don't use defaults. this is old style. Just use `frappe.db.get_single_value('Account Settings', 'time_sheet_for_sales_invoice')`

btw, why is this setting required?

File: erpnext/hr/doctype/process_payroll/process_payroll.js [hunk 3]
```diff
@@ -58,7 +58,23 @@ cur_frm.cscript.make_jv = function(doc, dt, dn) {
...
+	refresh: function(frm) {
+		// frm.trigger("toggle_fields")
+		frm.disable_save();
+	},
 
-frappe.ui.form.on("Process Payroll", "refresh", function(frm) {
-	frm.disable_save();
-});
+	// salary_slip_based_on_timesheet: function(frm) {
+// 		frm.trigger("toggle_fields")
```
Comment (rmehta) [inline erpnext/hr/doctype/process_payroll/process_payroll.js]: remove commented lines

File: erpnext/hr/doctype/salary_slip/salary_slip.py [hunk 4]
```diff
@@ -20,53 +20,103 @@ def autoname(self):
...
+				self.salary_slip_based_on_timesheet = ss_doc.salary_slip_based_on_timesheet or 0
+				self.set_time_sheet()
+				self.pull_sal_struct(ss_doc)
+
+	def set_time_sheet(self):
+		if self.salary_slip_based_on_timesheet and not self.get('timesheets'):
+			self.set("timesheets", [])
+
+			timesheets = frappe.db.sql(""" select * from `tabTime Sheet` where employee = %(employee)s and (status = 'Submitted' or
+				status = 'Billed') and %(start_date)s between start_date and end_date and %(end_date)s between start_date and end_date""",
```
Comment (rmehta) [inline erpnext/hr/doctype/salary_slip/salary_slip.py]: just check on basis of `docstatus` (?)

File: erpnext/hr/doctype/salary_slip/salary_slip.py [hunk 5]
```diff
@@ -20,53 +20,103 @@ def autoname(self):
...
+			self.salary_structure = ss_doc.name
+			self.hour_rate = ss_doc.hour_rate
+			self.total_working_hours = sum([d.working_hours or 0.0 for d in self.timesheets]) or 0.0
+			self.add_earning_for_hourly_wages(ss_doc.earning_type)
+
+	def add_earning_for_hourly_wages(self, earning_type):
+		default_type = False
+		for data in self.earnings:
+			if data.e_type == earning_type:
+				data.e_modified_amount = self.hour_rate * self.total_working_hours
```
Comment (rmehta) [inline erpnext/hr/doctype/salary_slip/salary_slip.py]: `e_modified_amount` (?) - fieldname is not descriptive

File: erpnext/projects/doctype/time_sheet/time_sheet.py [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+
+	def validate_mandatory_fields(self):
+		if self.production_order:
+			production_order = frappe.get_doc("Production Order", self.production_order)
+			pending_qty = flt(production_order.qty) - flt(production_order.produced_qty)
+
+		for data in self.timesheets:
+			if not data.from_time and not data.to_time:
+				frappe.throw(_("Row {0}: From Time and To Time is mandataory.").format(data.idx))
+
```
Comment (rmehta) [inline erpnext/projects/doctype/time_sheet/time_sheet.py]: `mandatory` spelling - do search and replace in full code

File: erpnext/accounts/doctype/accounts_settings/accounts_settings.py (hunk 2 above)
Comment (rohitwaghchaure) [inline erpnext/accounts/doctype/accounts_settings/accounts_settings.py]: Have require sys_defaults.time_sheet_for_sales_invoice in JS side and    
"depends_on": "eval:cint(sys_defaults.time_sheet_for_sales_invoice)" in JSON

File: erpnext/hr/doctype/salary_slip/salary_slip.py (hunk 5 above)
Comment (rohitwaghchaure) [inline erpnext/hr/doctype/salary_slip/salary_slip.py]: Can I change e_modified_amount to earning_amount and e_type to earning_type
d_type to deduction_type and d_modified_amount to deduction_amount

File: erpnext/accounts/doctype/accounts_settings/accounts_settings.json (hunk 1 above)
Comment (rohitwaghchaure) [inline erpnext/accounts/doctype/accounts_settings/accounts_settings.json]: OK

File: erpnext/hr/doctype/process_payroll/process_payroll.js (hunk 3 above)
Comment (rohitwaghchaure) [inline erpnext/hr/doctype/process_payroll/process_payroll.js]: Done

File: erpnext/hr/doctype/salary_slip/salary_slip.py (hunk 4 above)
Comment (rohitwaghchaure) [inline erpnext/hr/doctype/salary_slip/salary_slip.py]: Removed start and end date condition

File: erpnext/projects/doctype/time_sheet/time_sheet.py (hunk 6 above)
Comment (rohitwaghchaure) [inline erpnext/projects/doctype/time_sheet/time_sheet.py]: Done

File: erpnext/hr/doctype/salary_slip/salary_slip.py (hunk 5 above)
Comment (rmehta) [inline erpnext/hr/doctype/salary_slip/salary_slip.py]: :+1:

File: erpnext/accounts/doctype/accounts_settings/accounts_settings.py (hunk 2 above)
Comment (rmehta) [inline erpnext/accounts/doctype/accounts_settings/accounts_settings.py]: Load in as an __onload property in Sales Invoice.

Generally - its not very maintainable to add these to defaults, as it is not intuitive.

**Pull request merged**